// firebase-tools/lib/init/features/firestore/indexes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.initIndexes = exports.INDEXES_TEMPLATE = exports.DEFAULT_INDEXES_FILE = void 0;
const clc = require("colorette");
const error_1 = require("../../../error");
const api = require("../../../firestore/api");
const fsutils = require("../../../fsutils");
const prompt_1 = require("../../../prompt");
const logger_1 = require("../../../logger");
const templates_1 = require("../../../templates");
const indexes = new api.FirestoreApi();
exports.DEFAULT_INDEXES_FILE = "firestore.indexes.json";
exports.INDEXES_TEMPLATE = (0, templates_1.readTemplateSync)("init/firestore/firestore.indexes.json");
async function initIndexes(setup, config, info) {
    logger_1.logger.info();
    logger_1.logger.info("Firestore indexes allow you to perform complex queries while");
    logger_1.logger.info("maintaining performance that scales with the size of the result");
    logger_1.logger.info("set. You can keep index definitions in your project directory");
    logger_1.logger.info("and publish them with " + clc.bold("firebase deploy") + ".");
    logger_1.logger.info();
    info.indexesFilename =
        info.indexesFilename ||
            (await (0, prompt_1.input)({
                message: "What file should be used for Firestore indexes?",
                default: exports.DEFAULT_INDEXES_FILE,
            }));
    if (fsutils.fileExistsSync(info.indexesFilename)) {
        const msg = "File " +
            clc.bold(info.indexesFilename) +
            " already exists." +
            " Do you want to overwrite it with the Firestore Indexes from the Firebase Console?";
        if (!(await (0, prompt_1.confirm)(msg))) {
            info.writeIndexes = false;
            return;
        }
    }
    if (setup.projectId) {
        info.indexes = await getIndexesFromConsole(setup.projectId, info.databaseId);
    }
}
exports.initIndexes = initIndexes;
async function getIndexesFromConsole(projectId, databaseId) {
    const indexesPromise = indexes.listIndexes(projectId, databaseId);
    const fieldOverridesPromise = indexes.listFieldOverrides(projectId, databaseId);
    try {
        const res = await Promise.all([indexesPromise, fieldOverridesPromise]);
        return JSON.stringify(indexes.makeIndexSpec(res[0], res[1]), null, 2);
    }
    catch (e) {
        if (e.message.indexOf("is not a Cloud Firestore enabled project") >= 0) {
            return exports.INDEXES_TEMPLATE;
        }
        throw new error_1.FirebaseError("Error fetching Firestore indexes", {
            original: e,
        });
    }
}
