// firebase-tools/lib/mcp/tools/core/get_admin_sdk_config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.get_admin_sdk_config = void 0;
const zod_1 = require("zod");
const tool_js_1 = require("../../tool.js");
const util_js_1 = require("../../util.js");
const adminSdkConfig_js_1 = require("../../../emulator/adminSdkConfig.js");
exports.get_admin_sdk_config = (0, tool_js_1.tool)({
    name: "get_admin_sdk_config",
    description: "Gets the Admin SDK config for the current project. ",
    inputSchema: zod_1.z.object({}),
    annotations: {
        title: "Get Admin SDK Config",
        readOnlyHint: true,
    },
    _meta: {
        requiresAuth: true,
        requiresProject: true,
    },
}, async (_, { projectId }) => {
    const adminSdkConfig = await (0, adminSdkConfig_js_1.getProjectAdminSdkConfigOrCached)(projectId || "");
    if (!adminSdkConfig) {
        return (0, util_js_1.mcpError)(`No Admin SDK configuration found in project '${projectId || ""}'`);
    }
    return (0, util_js_1.toContent)(adminSdkConfig);
});
