// firebase-tools/lib/mcp/tools/core/list_apps.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.list_apps = void 0;
const zod_1 = require("zod");
const tool_js_1 = require("../../tool.js");
const util_js_1 = require("../../util.js");
const apps_js_1 = require("../../../management/apps.js");
exports.list_apps = (0, tool_js_1.tool)({
    name: "list_apps",
    description: "Retrieves apps registered in the current Firebase project.",
    inputSchema: zod_1.z.object({
        platform: zod_1.z
            .enum(["ios", "android", "web"])
            .nullish()
            .describe("the specific platform to list (omit to list all platforms)"),
    }),
    annotations: {
        title: "List Firebase Apps",
        readOnlyHint: true,
    },
    _meta: {
        requiresProject: true,
        requiresAuth: true,
    },
}, async ({ platform }, { projectId }) => {
    var _a;
    const apps = await (0, apps_js_1.listFirebaseApps)(projectId, (_a = platform === null || platform === void 0 ? void 0 : platform.toUpperCase()) !== null && _a !== void 0 ? _a : apps_js_1.AppPlatform.ANY);
    return (0, util_js_1.toContent)(apps);
});
